### PR TITLE
UHM-3614: Add logic to manage actions on visit page

### PR DESCRIPTION
**Pull request by mogoodrich**

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/mirebalais/apploader/CustomAppLoaderFactory.java [hunk 1]
```diff
@@ -2031,15 +2033,15 @@ private void enableCovid19() {
                 "link",
                 enterStandardHtmlFormLink(PihCoreUtil.getFormResource("covid19Intake.xml")),
                 Privileges.TASK_EMR_ENTER_COVID.privilege(),
-                and(sessionLocationHasTag(LocationTags.COVID_LOCATION))));
+                and(sessionLocationHasTag(LocationTags.COVID_LOCATION), visitDoesNotHaveEncounterOfType(EncounterTypes.COVID19_INTAKE))));
```
Comment (brandones) [inline api/src/main/java/org/openmrs/module/mirebalais/apploader/CustomAppLoaderFactory.java:2036]:
```suggestion
                and(sessionLocationHasTag(LocationTags.COVID_LOCATION),
                    visitDoesNotHaveEncounterOfType(EncounterTypes.COVID19_INTAKE))));
```

File: api/src/main/java/org/openmrs/module/mirebalais/apploader/CustomAppLoaderFactory.java [hunk 2]
```diff
@@ -2031,15 +2033,15 @@ private void enableCovid19() {
...
 
         extensions.add(visitAction(Extensions.COVID19_FOLLOWUP_VISIT_ACTION,
                 "ui.i18n.EncounterType.name." + EncounterTypes.COVID19_FOLLOWUP.uuid(),
                 "fas fa-fw fa-sun",
                 "link",
                 enterStandardHtmlFormLink(PihCoreUtil.getFormResource("covid19Followup.xml")),
                 Privileges.TASK_EMR_ENTER_COVID.privilege(),
-                and(sessionLocationHasTag(LocationTags.COVID_LOCATION))));
+                and(sessionLocationHasTag(LocationTags.COVID_LOCATION),visitHasEncounterOfType(EncounterTypes.COVID19_INTAKE))));
```
Comment (brandones) [inline api/src/main/java/org/openmrs/module/mirebalais/apploader/CustomAppLoaderFactory.java:2044]:
```suggestion
                and(sessionLocationHasTag(LocationTags.COVID_LOCATION),
                    visitHasEncounterOfType(EncounterTypes.COVID19_INTAKE))));
```

File: api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java [hunk 3]
```diff
@@ -35,6 +36,15 @@ public static String sessionLocationDoesNotHaveTag(LocationTagDescriptor descrip
         return new String("typeof sessionLocation !== 'undefined' && !hasMemberWithProperty(sessionLocation.tags, 'display','" + descriptor.name() + "')");
     };
 
+    // note that Java 8 Nashorn script engine support does not (fully?) support ECMAScript6 so we can't use arrow notation or map, etc
+    public static String visitHasEncounterOfType(EncounterTypeDescriptor descriptor) {
+        return new String("typeof visit !== 'undefined' && typeof visit.encounters !== 'undefined' && hasMemberThatEvaluatesTrue(visit.encounters, (function(encounter) { return encounter.encounterType.uuid === '" + descriptor.uuid() + "' } ))");
```
Comment (brandones) [inline api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java:41]: [Array.some()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) is ES5

Comment (brandones) [reply api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java:41]:
```suggestion
        return new String("visit && visit.encounters && " + 
          "visit.encounters.some(function(encounter) { " +
          "  return encounter.encounterType.uuid === '" + descriptor.uuid() +
          "' })");
```

File: api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java [hunk 4]
```diff
@@ -35,6 +36,15 @@ public static String sessionLocationDoesNotHaveTag(LocationTagDescriptor descrip
...
+    }
+
+    public static String visitDoesNotHaveEncounterOfType(EncounterTypeDescriptor descriptor) {
+        return new String("typeof visit !== 'undefined' && typeof visit.encounters === 'undefined' || !hasMemberThatEvaluatesTrue(visit.encounters, (function(encounter) { return encounter.encounterType.uuid === '" + descriptor.uuid() + "' } ))");
```
Comment (brandones) [inline api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java:45]:
```suggestion
        return new String("(visit && !visit.encounters) || " +
        "!visit.encounters.some(function(encounter) { " +
          "  return encounter.encounterType.uuid === '" + descriptor.uuid() +
          "' })");
```

Comment (mogoodrich) [reply api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java:48]: So I guess I'm not sure exactly what Nashorn supports, but it does not support "some" even if it is ECMA5... I basically tested a bunch of different methods in a unit test and nothing I wanted to use works.

I do agree with just changing the name of the method to "some" though... I was worried about naming collisions, but that's unlikely.  I will manually update the rest of this code and test.

Comment (brandones) [reply api/src/main/java/org/openmrs/module/mirebalais/require/RequireUtil.java:48]: Sounds good, thanks

**Pull request merged**